base/common/python/pki/nssdb.py

```diff
@@ -1241,8 +1241,15 @@ def __create_ski_ext(self, exts, ski_ext):
         if ski_ext.get('critical'):
             values.append('critical')
 
-        # generate subject key ID from hash (i.e. ignore the provided ID)
-        values.append('hash')
+        sk_id = ski_ext.get('sk_id')
+
+        if sk_id == 'DEFAULT':
+            # generate subject key ID from hash
+            values.append('hash')
+        else:
+            # convert <hex><hex>...<hex> into <hex>:<hex>:...:<hex>
+            value = binascii.unhexlify(sk_id).hex(':')
+            values.append(value)
```

Comment:
The value of `sk_id` (similar problem with the java `subjectKeyIdentifier`) can have two  meaning: an algorithm or an hex value. This difference is not explicit, is this approach used with other variables in the code?

Reply:
Actually you brought up a good point. In the Python code the extension is defined in a JSON object (since it's only used programmatically):
```
ski_ext = {
    'critical': ...,
    'sk_id': ...
}
```
The `sk_id` property could be either a `DEFAULT` or a `<hex><hex>...<hex>` since the value is loaded from [pki_req_ski](https://github.com/dogtagpki/pki/blob/master/docs/manuals/man5/pki_default.cfg.5.md?plain=1#L452-L455) param in `CS.cfg`. There's no param to specify the criticality so the `critical` property is unused right now.

The `NSSExtensionGenerator` is a relatively new mechanism to allow users to define the extensions, so the [configuration](https://github.com/dogtagpki/pki/wiki/PKI-NSS-Certificate-Extensions) is text-based which mimics [OpenSSL's configuration](https://www.openssl.org/docs/manmaster/man5/x509v3_config.html).

In OpenSSL's configuration the `subjectKeyIdentifier` seems to support only a single value: `none`, `hash`, or `<hex>:<hex>:...:<hex>`. It doesn't support `critical`, but other extensions in OpenSSL's configuration are multi-valued and do support `critical`.

My earlier patch was trying to do the same thing with `subjectKeyIdentifier` (except for the `none`), but since it's actually possible to specify the criticality, I've revised the patch to support multiple values including `critical`. Right now if there are multiple SKID values specified it will only use the last one. I think people who will use this will know that there can only be one SKID, so there's no need to check for multiple SKIDs, but we can add the check later if necessary.

Do you mind looking at the patch again? Thanks!

Reply:
I have not tested in OpenSSL but following the documentation `critical` is possible for all extensions including subjectKeyIdentifier so it is better to have.

Reply:
Ah yes, it says each param follows this format: `name = [critical, ]value(s)`. I haven't tried it myself either.